### PR TITLE
fix: Branch filtering not working

### LINKED ISSUE #4370: Branch filtering not working
**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior?**
The "Branches" drop down combo box does not work when more than one branch is listed (separated with spaces)

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem.**
If you write the name of one branch, the view is filtered to only show that branch. If you write more than one branch, only the first branch is shown. It also crashes when you try to edit the box with more than one branch names written on it.

**What is the expected behavior?**
Filter the history view showing all the written branches. Do not crash when editing the field.

**Environment you encounter the issue:**
- GitExtensions version: 2.51.RC2
- GIT version: 2.15.1.2
- OS version: Both Win7 x64 and Win10 1703

**Did this work in previous version of GitExtensions (which)?**
Yes. 2.50.02

### PR DESCRIPTION
**Pull request by RussKie**
The "Branches" drop down combo box does not work when more than one branch is listed (separated with spaces).
The error occured while attempting to read the text from the dropdown when it had no items in the list.

Fixes #4370

What did I do to test the code and ensure quality:
 - ran it manually in various scenarios comparing with 2.50 version

Steps to reproduce:
* type "master" and hit Enter - this filters the branch list to master branch only
* add one or two spaces after "master" -> "master  " and hit Enter - at this stage I typically get AOORE because the dropdown has no items in it (all filtered out) and we are accessing `dropdown.Text` property which throws the exception

Has been tested on (remove any that don't apply):
 - Windows 8.1

### CODE REVIEW

File: GitUI/FilterBranchHelper.cs [hunk 1]
```diff
@@ -8,6 +8,7 @@ namespace GitUI
 {
     public class FilterBranchHelper : IDisposable
     {
+        private bool _applyingFilter;
```
Comment (gerhardol) [inline GitUI/FilterBranchHelper.cs]: Init?

File: GitUI/FilterBranchHelper.cs [hunk 2]
```diff
@@ -143,18 +144,35 @@ private void toolStripBranches_DropDown(object sender, EventArgs e)
 
         private void ApplyBranchFilter(bool refresh)
         {
-            bool success = _NO_TRANSLATE_RevisionGrid.SetAndApplyBranchFilter(_NO_TRANSLATE_toolStripBranches.Text);
-            if (success && refresh)
-                _NO_TRANSLATE_RevisionGrid.ForceRefreshRevisions();
+            if (_applyingFilter)
```
Comment (gerhardol) [inline GitUI/FilterBranchHelper.cs]: Not fully threadsafe, but should be good enough

File: GitUI/FilterBranchHelper.cs (hunk 1 above)
Comment (RussKie) [reply GitUI/FilterBranchHelper.cs]: fields are automatically initialised; ref types to `null`, value types to their default value (in this case it is `false`m which is we want)

File: GitUI/FilterBranchHelper.cs (hunk 2 above)
Comment (RussKie) [reply GitUI/FilterBranchHelper.cs]: I thought of it, but the helper is used from FormBrowse on the UI thread, so there can't be any x-treading.

**Comment by RussKie** [comment]
I seem to have consistently repro'd it in the following way:
* type "master" and hit Enter - this filters the branch list to master branch only
* add one or two spaces after "master" -> "master  " and hit Enter - at this stage I typically get AOORE because the dropdown has no items in it (all filtered out) and we are accessing `dropdown.Text` property which throws the exception

**Comment by gerhardol** [comment]
> add one or two spaces after "master" -> "master " and hit Enter - at this stage I typically get AOORE 

Yes, this was what I tried and I do not see the exception and it is OK typing the second branch and get filtering.
It does not matter if you could reproduce.

**Comment by RussKie** [comment]
It is somewhat intermittent - sometimes you get a constant barrage of them, sometimes not...

* Type "master", hit Enter
* Add " " -> "master ", hit Enter
* Add " " and get AOORE

**Comment by RussKie** [comment]
I'm standing by to give the OP to test the fix.
If he doesn't respond in few days, I'll get it merged and hopefully we can finally get 2.51 out

**Pull request merged**